README.md

```diff
@@ -312,7 +345,23 @@ for (const key of map.keys()) {
 // key: {"b":2}
 ```
 
-Remember to be cautious with use of powerful constructs like this iterable WeakMap. Web APIs designed with semantics analogous to these are widely considered to be legacy mistakes. It’s best to avoid exposing garbage collection timing in your applications, and to use weak references and finalizers sparingly, e.g., for cases which help reduce memory usage.
+Remember to be cautious with use of powerful constructs like this iterable WeakMap. Web APIs designed with semantics analogous to these are widely considered to be legacy mistakes. It’s best to avoid exposing garbage collection timing in your applications, and to use weak references and finalizers only where a problem cannot be reasonably solved in other ways.
+
+### Finalizers, visibility, and turns: final notes of caution
+
+As mentioned above, finalizers run after an object is collected, in a separate microtask.  An object can be collected at any point where it is no longer needed to compute the result of a computation.  JavaScript implementations have a wide latitude to determine what this means in practice.  For example, consider this async function:
+
+```js
+async function process(x) {
+  const p = x.getPtr();
+  await new Promise(resolve => setTimeout(resolve));
+  return handle(p);
+}
+```
+
+If `x` has a finalizer that invalidates the value returned by `getPtr()`, an implementation is free to collect `x` at any point after the `getPtr()` call, and to finalize it during the `await`, making the `handle(p)` call operate on an invalid pointer.  From a language perspective, the fact that `x` was an argument to a function does not prevent it from being collected.  Additionally the `return handle(p)` call is a tail call, which an engine may implement as throwing away any stack frame with the `x` binding, which is another opportunity for `x` to become collectible.
```

Comment:
> invalidates the value returned by

Maybe say invalidates the object?
The Javascript reference `p` itself will stay valid of course, it's the object that is referenced that may be internally invalidated by library logic. So a primitive value can't be invalidated

Reply:
Maybe somethings like this:

> In this example, an implementation is free to collect `x` at any point after the `getPtr()` call. From a language perspective, the fact that `x` was an argument to a function does not prevent it from being collected.\
If `x` has a finalizer that invalidates the object returned by `getPtr()`, it may be called during the `await`, making the `handle(p)` call operate on an invalid object.

Reply:
What is `getPtr`?

Reply:
I'm assuming a method that return a wrapper "pointer" object to `x`, that can be invalidated?

Reply:
Sorry, I think the confusion is from the name `getPtr()`.  I meant to return a member field of `x` which, when `x`'s finalizer runs, becomes no longer valid; e.g. perhaps `x`'s finalizer frees webassembly memory pointed to by `p`.  I think the name is confusing in that it suggests that you retrieve a pointer to `x`, but that's not what i meant.

An example of this pattern is the `$$` object used in emscripten's embind.